src/main/java/org/opensearch/security/auditlog/sink/InternalOpenSearchDataStreamSink.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.security.auditlog.sink;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+
+import org.opensearch.action.index.IndexRequestBuilder;
+import org.opensearch.action.DocWriteRequest;
+import org.opensearch.action.admin.indices.datastream.CreateDataStreamAction;
+import org.opensearch.action.admin.indices.template.put.PutComposableIndexTemplateAction;
+import org.opensearch.action.support.master.AcknowledgedResponse;
+import org.opensearch.action.support.WriteRequest.RefreshPolicy;
+import org.opensearch.action.admin.indices.datastream.GetDataStreamAction;
+import org.opensearch.cluster.metadata.ComposableIndexTemplate;
+import org.opensearch.client.Client;
+import org.opensearch.client.IndicesClient;
+import org.opensearch.client.indices.CreateDataStreamRequest;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.common.util.concurrent.ThreadContext.StoredContext;
+import org.opensearch.common.xcontent.XContentHelper;
+import org.opensearch.security.auditlog.impl.AuditMessage;
+import org.opensearch.security.support.ConfigConstants;
+import org.opensearch.security.support.HeaderHelper;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.core.xcontent.MediaTypeRegistry;
+import org.opensearch.core.xcontent.XContentParser;
+import org.opensearch.core.common.bytes.BytesArray;
+import org.opensearch.core.xcontent.NamedXContentRegistry;
+import org.opensearch.ResourceAlreadyExistsException;
+import org.opensearch.transport.RemoteTransportException;
+
+import org.joda.time.format.DateTimeFormat;
+import org.joda.time.format.DateTimeFormatter;
+
+public final class InternalOpenSearchDataStreamSink extends AuditLogSink {
```

Comment:
There may be an opportunity to abstract the logic that is common to here and InternalOpenSearchSink into an Abstract class that both classes can inherit. These classes are pretty small, so it won't reduce the code too much but I thought I'd leave a comment  none-the-less.

Reply:
I created an abstracted logic with AbstractInternalOpenSearchSink and shared duplicate code in this class.